### ISSUE TITLE
Generated OTP not equal to that of other software

### ISSUE THREAD

**Issue by SolarLiner** [bilelmoussaoui/Authenticator#80]
**Reporter info**

 ```
 * Distribution - Ubuntu 18.04
 * Gtk+ 3.0 version - 3.22.30
 * Desktop environment - GNOME
 ```

**Authenticator version**: 0.2.2.1

**Actual issue**
Generated OTP using the same key in Authenticator and other software (in my case both Google Authenticator and KeePassXC) does not yield the same results, which would not make the Authenticator-generated OTP usable.

**Steps to reproduce (if you knew)**

* Add provider in Authenticator and in other OTP generator apps at the same time
* Watch OTP generated from both Authenticator and another app proven to work - they differ

![capture d ecran de 2018-09-04 22-32-14](https://user-images.githubusercontent.com/4302963/45056718-fd3b4080-b093-11e8-99de-7823ef7cf7a4.png)

**Other Note (feature-request, question, etc...)**
Not knowing a lot about 2FA I still tested to see if Authenticator was still getting out valid OTPs, just in case; but testing on Twitter doesn't let me log in.
Also, and while this might not be the best place to ask for it, but generally the timeouts aren't the same across apps, but the codes are the same anyways, and still work as expected. How does *that* work?

**Comment by dannosaur** [comment]
Might it be a case that Authenticator is displaying a previous code as opposed to a wrong one?

Look at your OTP that's working and make a note of the code, then wait for the codes to cycle and see if Authenticator at any point in time displays the correct code. You might be on to something with the timing issue (I've see a difference in codes on mine as well, that's caused by the code expiring at differing times).

If this is the case, I can see why this is happening.

**Comment by bilelmoussaoui** [comment]
I totally agree with @dannosaur. Currently, I assume that all codes will be  "out of date" in 30 seconds. It was on of the reasons to keep the database, in order to have some sort of which providers use 30 seconds or 60 seconds before the code is out of date.

**Comment by dannosaur** [comment]
@bilelmoussaoui upon reading up on TOTP, the delta is usually keyed to the system time, and changes whenever T=0 or T=30 epoch (in a 30 second interval code). As it seems right now it's just counting down from 30 when the application starts, so if you start it at the wrong time, all of your codes, in theory, will be incorrect all the time.

**Comment by SolarLiner** [comment]
Tracking codes as they're generated from both KeePassXC and Authenticator doesn't reveal any cycle; here's generated TOTPs taken over the span of 3 minutes:

| KeePassXC TOTP | Authenticator TOTP |
|:----------------:|:-------------------:|
|     856 607    |       016 902      |
|     575 631    |       737 399      |
|     836 755    |       600 194      |
|     689 976    |       343 052      |
|     204 101    |       702 254      |

Note that there only was a couple of seconds of delay between the two apps' cycles, and none of the KeePass OTP show in the Authenticator, even after in a later (or earlier) cycle.

**Comment by dannosaur** [comment]
That's really interesting. And confusing. I've seen this problem with the codes on screen not matching my phone's authenticator app, but the codes eventually syncronized with the "correct" codes, even though they were out of sync.

This looks more to me like the secret is just wrong. How did you get the secret into Authenticator? Did you copy/paste or type it in, or did you scan a QR code?

**Comment by bilelmoussaoui** [comment]
@SolarLiner The only difference between both versions didn't touch the code related to 2FA. It uses pyotp, with the same library version as before.
How does KeePass alllows you to copy the token??? I'm not sure if it's a valid token that KeePass allows you to copy. You had this issue with which website? So I can try to see if I can reproduce it

**Comment by SolarLiner** [comment]
Yes, KeePassXC stores the secret as a custom entry field in the encrypted password file, which allows me to access it once decrypted.

All accounts where I use 2FA are subject to this bug, so in my case that includes Discord, Twitter and Google (through YouTube, but honestly I don't know if I took the right one).

Maybe the bug has to do with `pyotp` and maybe even system dependencies. An unlikely scenario considering I'm running a fresh Ubuntu 18.04 install (I had just installed it back and Authenticator was one of the first apps I installed once I got flatpak!), so that would be unlikely. I'll be spinning VMs to investigate that.

**Comment by SolarLiner** [comment]
Reporting from testing in Ubuntu 18.10; the app indeed works as expected (wrt. this issue, as I've found another bug tied to the New Account dialog where the account name and secret input boxes' values are swapped, leading to the secret being displayed as the account name and the app not being able to generate OTPs), and shows the same OTP as in KeePassXC.

From Ubuntu 18.04, however, the application freezes when trying to add an account. I solved that issue by logging out and logging back in, as I had the "automatic login on boot" option enabled which doesn't unlock the keyring. After that, adding a new account behaved normally and didn't feature the bug found in Ubuntu 18.10. However, I could reproduce the initial bug described by this issue: the generated OTPs aren't the same ones generated by KeePassXC. As an extra measure, I also verified that both the VM Authenticator app and the one on the host machine were generating the same OTPs, which they didn't; and while doing that the same new account dialog bug I found in Ubuntu 18.10 started appearing on my host machine version, strangely!

Finally, testing in Solus yields a perfectly working application, without the Ubuntu 18.10 bug, and working OTP (generating + correct).

**Comment by bilelmoussaoui** [comment]
It must be an issue related to something else other than the application itself. It's the reason why I'm distributing the application as Flatpak only. This way, i'm pretty sure that everything works as expected.

The latest release on Flathub has all those small issues you have fixed. Can you try it and let me know if it has any issue of the ones you listed?

**Comment by SolarLiner** [comment]
I had totally forgotten that flatpak bundles dependencies along with the app... I have used Flatpak to install Authenticator on all three VMs, so in theory, the result should have indeed been the same no matter the OS or system dependencies, yet the differences *are* there.

I'll be checking for updates on my physical machine and report back.

**Comment by SolarLiner** [comment]
Actually, is the "latest release" 0.2.2.1? If so, that is the version with which I tested the app on all 3 VMs and currently have on my physical machine.

**Comment by bilelmoussaoui** [comment]
Can you guys try the 0.2.3 stable version and let me know if you still have this issue? I would have to investigate a bit more to see how I can reproduce this and fix it.

**Comment by SolarLiner** [comment]
Built the flatpak from the git tag as flatpak *really* didn't believe there was an update available.

**It works**!!! Also getting to see the redesign, which I like quite a lot to be honest. But that's a topic for another issue.
